### PR TITLE
Controller wrap

### PR DESCRIPTION
**Pull request by m-kostrzewa**
- wraps needed controller client's functionality
- unit tests
- change proj structure

### CODE REVIEW

File: driver/controller.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+func (c *Controller) GetDefaultGatewayIp(net *types.VirtualNetwork) (string, error) {
+	ipamReferences, err := net.GetNetworkIpamRefs()
+	if err != nil {
+		return "", err
+	}
+	if len(ipamReferences) == 0 {
+		return "", errors.New("Ipam references list is empty")
+	}
+	attribute := ipamReferences[0].Attr
```
Comment (Attil) [inline driver/controller.go:46]: Why 1st element?

File: driver/controller.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	ipamSubnets := attribute.(types.VnSubnetsType).IpamSubnets
+	if len(ipamSubnets) == 0 {
+		return "", errors.New("Ipam subnets list is empty")
+	}
+	gw := ipamSubnets[0].DefaultGateway
```
Comment (Attil) [inline driver/controller.go:51]: Again, 1st element.

File: driver/controller.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	if gw == "" {
+		return "", errors.New("Default GW is empty")
+	}
+	return gw, nil
+}
+
+func (c *Controller) GetOrCreateInstance(tenantName, containerId string) (*types.VirtualMachine, error) {
+	name := fmt.Sprintf("%s:%s:%s", DomainName, tenantName, containerId)
+	instance, err := types.VirtualMachineByName(c.ApiClient, name)
```
Comment (Attil) [inline driver/controller.go:60]: VirtualMachine? I assume there's no ContainerByName?

File: driver/controller.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	return iface, nil
+}
+
+func (c *Controller) GetInterfaceMac(iface *types.VirtualMachineInterface) (string, error) {
+	macs := iface.GetVirtualMachineInterfaceMacAddresses()
+	if len(macs.MacAddress) == 0 {
+		return "", errors.New("Empty MAC list")
+	}
+	return macs.MacAddress[0], nil
```
Comment (Attil) [inline driver/controller.go:106]: Again, first

File: driver/controller.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func (c *Controller) GetOrCreateInstanceIp(net *types.VirtualNetwork,
+	iface *types.VirtualMachineInterface) (*types.InstanceIp, error) {
+	ifaceFQName := iface.GetFQName()
+	tenantName := ifaceFQName[len(ifaceFQName)-2]
```
Comment (Attil) [inline driver/controller.go:112]: What? Second last?

File: tests/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+	tenantName   = "agatka"
+	networkName  = "test_net"
+	subnetPrefix = "10.10.10.0"
+	subnetMask   = 24
+	defaultGW    = "10.10.10.1"
+	ifaceMac     = "contrail_pls_check_macs"
+	containerID  = "12345678901"
+)
+
+var _ = Describe("Controller", func() {
```
Comment (Attil) [inline tests/controller_test.go:34]: All suggestions I've mentioned in the other pull and...

File: tests/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+	err := c.Create(testInstance)
+	Expect(err).ToNot(HaveOccurred())
+	return testInstance
+}
+
+func createMockedInterface(c contrail.ApiClient, instance *types.VirtualMachine,
+	net *types.VirtualNetwork) *types.VirtualMachineInterface {
+	iface := new(types.VirtualMachineInterface)
+	instanceFQName := instance.GetFQName()
+	namespace := instanceFQName[len(instanceFQName)-2]
```
Comment (Attil) [inline tests/controller_test.go:296]: 2nd last character?

**Review by Attil — COMMENTED** [review]
I'll post some more when I debug this.

File: driver/controller.go (hunk 3 above)
Comment (m-kostrzewa) [reply driver/controller.go:60]: nah, Contrail doesn't have container data type.

File: tests/controller_test.go (hunk 7 above)
Comment (m-kostrzewa) [reply tests/controller_test.go:296]: instanceFQName is an array of strings, this means second to last string
I'm not sure how to refactor this without complicating more than it's worth

File: driver/controller.go (hunk 4 above)
Comment (m-kostrzewa) [reply driver/controller.go:106]: I'm not sure which array element to take yet, maybe figure it out in another PR?

**Comment by m-kostrzewa** [comment]
applied CR remarks

**Pull request closed without merging**